annotated.go

```diff
@@ -126,6 +130,66 @@ func newAnnotations(fType reflect.Type) annotations {
 	}
 }
 
+// field used for embedding fx.Out type in generated struct.
+var _outAnnotationField = reflect.StructField{
+	Name:      "Out",
+	Type:      reflect.TypeOf(Out{}),
+	Anonymous: true,
+}
+
+// This generates the annotated Out struct when function is annotated
+// with fx.As and/or fx.ResultTags options.
+func (ann *annotations) genAnnotatedOutStruct() error {
+	if !ann.annotatedOut && !ann.annotatedAs {
+		return nil
+	}
+	fType := ann.fType
+	offsets := make([]int, fType.NumOut())
+	annotatedResult := []reflect.StructField{_outAnnotationField}
+	for i := 0; i < fType.NumOut(); i++ {
+		if fType.Out(i) == _typeOfError {
+			ann.returnsError = true
+			continue
+		}
+		structFieldType, err := ann.structFieldType(i)
+		if err != nil {
+			return err
+		}
+		structField := genAnnotatedOutStructField(i, structFieldType)
+		if i < len(ann.outTags) {
+			structField.Tag = reflect.StructTag(ann.outTags[i])
+		}
+		offsets[i] = len(annotatedResult)
+		annotatedResult = append(annotatedResult, structField)
+	}
+	ann.Outs = []reflect.Type{reflect.StructOf(annotatedResult)}
+	ann.resultOffsets = offsets
+	if ann.returnsError {
+		ann.Outs = append(ann.Outs, _typeOfError)
+	}
+	return nil
+}
+
+// helper for getting type of a fx.Out struct field
+func (ann *annotations) structFieldType(i int) (reflect.Type, error) {
+	if ann.annotatedAs {
+		asType := reflect.TypeOf(ann.asTargets[i]).Elem()
```

Comment:
Should we turn the asTargets into the destination type in asAnnotation.apply?
That would make it clear that ann.asTargets is the actual target type.